driver-core/src/main/com/mongodb/internal/connection/LoadBalancedCluster.java

```diff
@@ -181,50 +202,211 @@ public ServerTuple selectServer(final ServerSelector serverSelector) {
+                    settings.getServerSelectionTimeout(MILLISECONDS), settings.getSrvHost()));
+        } else {
+            return new MongoTimeoutException(format("Timed out after %d ms while waiting to resolve SRV records for %s. "
+                            + "Resolution exception was '%s'",
+                    settings.getServerSelectionTimeout(MILLISECONDS), settings.getSrvHost(), localSrvResolutionException));
+        }
+    }
+
+    private long getMaxWaitTimeNanos() {
+        if (settings.getServerSelectionTimeout(NANOSECONDS) < 0) {
+            return Long.MAX_VALUE;
+        }
+        return settings.getServerSelectionTimeout(NANOSECONDS);
+    }
+
+    private void notifyWaitQueueHandler(final ServerSelectionRequest request) {
+        lock.lock();
+        try {
+            if (isClosed()) {
+                request.onError(createShutdownException());
+                return;
+            }
+            if (initializationCompleted) {
+                handleServerSelectionRequest(request);
+                return;
+            }
+
+            waitQueue.add(request);
+
+            if (waitQueueHandler == null) {
+                waitQueueHandler = new Thread(new WaitQueueHandler(), "cluster-" + clusterId.getValue());
+                waitQueueHandler.setDaemon(true);
+                waitQueueHandler.start();
+            } else {
+                condition.signalAll();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private final class WaitQueueHandler implements Runnable {
+        public void run() {
+            List<ServerSelectionRequest> timeoutList = new ArrayList<>();
+            while (!(isClosed() || initializationCompleted)) {
+                lock.lock();
+                try {
+                    if (isClosed() || initializationCompleted) {
+                        break;
+                    }
+                    long waitTimeNanos = Long.MAX_VALUE;
+                    long curTimeNanos = System.nanoTime();
+
+                    for (Iterator<ServerSelectionRequest> iterator = waitQueue.iterator(); iterator.hasNext();) {
+                        ServerSelectionRequest next = iterator.next();
+                        long remainingTime = next.getRemainingTime(curTimeNanos);
+                        if (remainingTime <= 0) {
+                            timeoutList.add(next);
+                            iterator.remove();
+                        } else {
+                            waitTimeNanos = Math.min(remainingTime, waitTimeNanos);
+                        }
+                    }
+                    if (timeoutList.isEmpty()) {
+                        try {
+                            //noinspection ResultOfMethodCallIgnored
+                            condition.await(waitTimeNanos, NANOSECONDS);
+                        } catch (InterruptedException e) {
+                            fail();
```

Comment:
I remember that `fail()` was added as an alternative to `assertTrue(false`), but I so not remember why `throw new AssertionError()` was discarded (was it even considered?). @jyemin do you remember?

Reply:
I don't recall, but perhaps just centralization of all AssertionError creation to a single class.